crates/rethnet_evm/src/state/fork.rs

```diff
@@ -1,219 +1,154 @@
-            latest_state_root: RwLock::new(generated_state_root),
-            state_root_to_block_number: HashMap::new(),
-        }
-    }
+        let mut local_state = HybridState::with_accounts(accounts);
+        local_state.checkpoint().unwrap();
 
-    /// if not already cached, fetches from remote then caches; then returns cached
-    fn get_remote_account_info(
-        &self,
-        address: &Address,
-        block_number: U256,
-    ) -> Result<Option<AccountInfo>, super::StateError> {
-        let account_info_cache = self.account_info_cache.upgradable_read();
-        if let Some(cached) = account_info_cache.get(&(*address, block_number)) {
-            Ok(Some(cached.clone()))
-        } else if let Some(mut remote_account) = self.remote_state.basic(*address)? {
-            let mut account_info_cache = RwLockUpgradableReadGuard::upgrade(account_info_cache);
-            account_info_cache.insert((*address, block_number), remote_account.clone());
-
-            if let Some(code) = remote_account.code.take() {
-                let mut code_by_hash_cache = self.code_by_hash_cache.write();
-                code_by_hash_cache.insert(remote_account.code_hash, code);
-            }
+        let generated_state_root = hash_generator.lock().next_value();
+        let mut state_root_to_state = HashMap::new();
+        let local_root = local_state.state_root().unwrap();
+        state_root_to_state.insert(generated_state_root, local_root);
 
-            Ok(Some(remote_account))
-        } else {
-            Ok(None)
+        Self {
+            local_state,
+            remote_state: Arc::new(Mutex::new(CachedRemoteState::new(remote_state))),
+            fork_block_number,
+            state_root_to_state: RwLock::new(state_root_to_state),
+            current_state: RwLock::new((generated_state_root, local_root)),
+            initial_state_root: generated_state_root,
+            hash_generator,
         }
     }
 }
 
 impl StateRef for ForkState {
-    type Error = super::StateError;
+    type Error = StateError;
 
     fn basic(&self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
-        let latest_state_root = *self.latest_state_root.read();
-
-        let block_number = self
-            .state_root_to_block_number
-            .get(&latest_state_root)
-            .unwrap_or(&self.fork_block_number);
-
-        if block_number < &self.fork_block_number {
-            self.get_remote_account_info(&address, *block_number)
-        } else if let Some(layered) = self.layered_state.basic(address)? {
-            Ok(Some(layered))
-        } else if let Some(remote) = self.get_remote_account_info(&address, *block_number)? {
-            Ok(Some(remote))
+        if let Some(local) = self.local_state.basic(address)? {
+            Ok(Some(local))
         } else {
-            Ok(None)
+            self.remote_state.lock().basic(address)
         }
     }
```

Comment:
Yes, in that case the `local_state` will be set to the initial state and result in no result found. Thus it will always fall back to the `else` statement